src/ServiceControl.Persistence.RavenDb5/RavenAttachmentsBodyStorage.cs

```diff
@@ -1,43 +1,61 @@
 ﻿namespace ServiceControl.Operations.BodyStorage.RavenAttachments
 {
+    using System;
     using System.IO;
+    using System.Linq;
     using System.Threading.Tasks;
     using Raven.Client.Documents;
-    using Raven.Client.Documents.Commands.Batches;
-    using Sparrow.Json.Parsing;
+    using Raven.Client.Documents.Session;
+    using ServiceControl.MessageFailures;
+    using ServiceControl.MessageFailures.Api;
 
     class RavenAttachmentsBodyStorage : IBodyStorage
     {
-        const string AttachmentName = "body";
+        public const string AttachmentName = "body";
         readonly IDocumentStore documentStore;
 
         public RavenAttachmentsBodyStorage(IDocumentStore documentStore)
         {
             this.documentStore = documentStore;
         }
 
-        // TODO: This method is only used in tests and not by ServiceControl itself! But in the Raven3.5 persister, it IS used!
-        // It should probably be removed and tests should use the RavenDbRecoverabilityIngestionUnitOfWork
-        public async Task Store(string messageId, string contentType, int bodySize, Stream bodyStream)
+        public Task Store(string uniqueId, string contentType, int bodySize, Stream bodyStream)
+            => throw new NotImplementedException("Only included for interface compatibility with Raven3.5 persister implementation. Raven5 tests should use IIngestionUnitOfWorkFactory to store failed messages/bodies.");
+
+        public async Task<MessageBodyStreamResult> TryFetch(string bodyId)
         {
-            var documentId = MessageBodyIdGenerator.MakeDocumentId(messageId);
+            using var session = documentStore.OpenAsyncSession();
+
+            // BodyId could be a MessageID or a UniqueID, but if a UniqueID then it will be a DeterministicGuid of MessageID and endpoint name and be Guid-parseable
+            // This is preferred, then we know we're getting the correct message body that is attached to the FailedMessage document
+            if (Guid.TryParse(bodyId, out _))
```

Comment:
More by the API history than what Core is forcing through our abstraction.

The reason for a Guid.TryParse() is that a uniqueID is always going to be built by DeterministicGuid and will always be parseable as a Guid. This might be true of a MessageID too, but if you had a custom message ID strategy or you had an MSMQ-like `{Guid}\{int}` construction, then we can easily skip this part knowing it's not going to succeed.